### DISCUSSION TITLE
How create _inc.tsp for dummies

### DISCUSSION THREAD

**Discussion by flexguse** [john30/ebusd#1410] [Q&A]
Hi everybody,

I need some help (for dummies!) creating custom _inc.tsp configurations to read values which are not provided by default.

First of all I would like to give you some information about my Vaillant heating.

```bash
root@d310c0f97dcf:/# ebusctl grab result
1008b5110100 / 08fe011226040f0001 = 59
10feb510020601 = 12
1008b5120204ff / 0101 = 12
1008b513020508 / 00 = 12
1008b5100305ff01 / 0101 = 12
1052b5030c0700ffffffffffffffffffff / 0101 = 4

root@d310c0f97dcf:/# ebusctl scan result
08;Vaillant;BAI00;0104;7803;21;18;47;0010021924;0001;012018;N6
15;Vaillant;70000;0510;6403;21;18;52;0020242192;0082;084976;N6
52;Vaillant;VR_70;0109;2903;??;??;??;???0184843;0082;016060;N4

root@d310c0f97dcf:/# ebusctl info
version: ebusd 24.1.24.1
update check: OK, device firmware v5 1[4a1b] available
device: 192.168.178.59:9999, TCP, enhanced, firmware 1.1[4a08].1[4a08]
signal: acquired
symbol rate: 38
max symbol rate: 155
min arbitration micros: 1
max arbitration micros: 44
min symbol latency: 6
max symbol latency: 46
scan: finished
reconnects: 0
masters: 3
messages: 700
conditional: 3
poll: 0
update: 10
address 03: master #11
address 08: slave #11, scanned "MF=Vaillant;ID=BAI00;SW=0104;HW=7803", loaded "vaillant/bai.308523.inc", "vaillant/08.bai.csv"
address 10: master #2
address 15: slave #2, scanned "MF=Vaillant;ID=70000;SW=0510;HW=6403", loaded "vaillant/15.700.csv"
address 31: master #8, ebusd
address 36: slave #8, ebusd
address 52: slave, scanned "MF=Vaillant;ID=VR_70;SW=0109;HW=2903", loaded "vaillant/52.vr_70.csv"
``` 
I've been successful reading several data and putting it via MQTT into a Influx DB and creating a custom Grafana Dashboard for more than a year now.

Now I would like to read some more attributes (i.e. d.29 - primary circuit water flow rate) from my heating to optimize the heating settings but PrimaryCircuitFlowRate_DK is not provided for my heating model. When I try to read the value this happens:

```bash
root@d310c0f97dcf:/# ebusctl read PrimaryCircuitFlowRate_DK
ERR: element not found
```

which is clear, because `bai.308523.inc` does not include this attribute.

As a next step I used devpod to create a VS workspace using the .devcontainer settings. This works fine, I'm able to use ebus-notebook and my ebusd instance is connected.

After that I searched the configurations for d.29 and created a custom bai.0010021924_inc.tsp file:

```
import "@ebusd/ebus-typespec";
import "./_templates.tsp";
using Ebus;
using Ebus.Num;

namespace Vaillant;

namespace Bai._0010021924_inc{

  /** d.29 PrimaryCircuitFlowrate_DK */
  @inherit(r)
  @ext(0xfb, 0)
  model PrimaryCircuitFlowrate {
    /** value of flowrate */
    @unit("l/h")
    value: UIN;
  }
}
```

When I run `Convert model to eBUS CSV and test...` the result is

```bash
conversion result:
type,circuit,level,name,comment,qq,zz,pbsb,id,*name,part,type,divisor/values,unit,comment
*r,
r,Main,,PrimaryCircuitFlowrate,d.29 PrimaryCircuitFlowrate_DK,,,b509,0dfb00,value,,UIN,,l/h,value of flowrate
ebusd result:
read -V -V -def -def -i "" "r,Main,PrimaryCircuitFlowrate,d.29 PrimaryCircuitFlowrate_DK,,,b509,0dfb00,value,,UIN,,l/h,value of flowrate"
error: AggregateError
```

Now I'm lost.

1. How do I find out at which address I can read the value for d.29? Step by step guidiance is highly appreciated.
2. What is the best way to find out how to convert the return value into something human readable?

**Comment by JonesGW-MGD** [comment]
I don't know whether this helps.  I scoured a few config files about a year ago for d.29 and found this one  - (which looks like the same PBSB code as you have).  I have an Ecotec Exclusive IQ 835 (at leas that is what it is called in the UK)

`r,,d.29_PrimaryCircuitFlowRate,d.29 PrimaryCircuitFlowRate_DK,,,,"FB00",,,uin100,,,primary circuit water flow rate m3/h`

converted to run in -def mode, does this give you anything?  (My circuit is bai and address 08):

`ebusctl read -f -def "r,bai,d.29_PrimaryCircuitFlowRate,PrimaryCircuitFlowRate_DK,,08,b509,0dfb00,uin100,s,UIN,100,Primary Circuit Water Flow Rate"`

It gives me a value - I'm just not sure it seems right - it's hard for me to check d.29 on the installers panel at the same time! The units in the description are suspect - I think instead it is l/h/100 and not l/h/1000 which is what l/h to m3/h should be I think?!  But the flow rate feels higher that I expect it to be - just now it is showing as '15.1' and my panel checked a min later shows about 850l/h so your post has prompted me do some more digging at least!

(Sorry, I'm not competent enough to triage why yours errors - maybe you are missing a default)
